RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletHeterogeneousEDProducer.cc

```diff
@@ -0,0 +1,206 @@
+#include "HeterogeneousCore/CUDAServices/interface/CUDAService.h"
+#include "HeterogeneousCore/Producer/interface/HeterogeneousEDProducer.h"
+
+#include "CAHitQuadrupletGenerator.h"
+#include "CAHitQuadrupletGeneratorGPU.h"
+#include "RecoPixelVertexing/PixelTriplets/interface/OrderedHitSeeds.h"
+#include "RecoTracker/TkHitPairs/interface/IntermediateHitDoublets.h"
+#include "RecoTracker/TkHitPairs/interface/RegionsSeedingHitSets.h"
+
+namespace {
+void fillNtuplets(RegionsSeedingHitSets::RegionFiller &seedingHitSetsFiller,
+                  const OrderedHitSeeds &quadruplets) {
+  for (const auto &quad : quadruplets) {
+    seedingHitSetsFiller.emplace_back(quad[0], quad[1], quad[2], quad[3]);
+  }
+}
+} // namespace
+
+class CAHitNtupletHeterogeneousEDProducer
+    : public HeterogeneousEDProducer<heterogeneous::HeterogeneousDevices<
+          heterogeneous::GPUCuda, heterogeneous::CPU>> {
+public:
+  CAHitNtupletHeterogeneousEDProducer(const edm::ParameterSet &iConfig);
+  ~CAHitNtupletHeterogeneousEDProducer() = default;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions &descriptions);
+  void beginStreamGPUCuda(edm::StreamID streamId,
+                          cuda::stream_t<> &cudaStream) override;
+
+  void acquireGPUCuda(const edm::HeterogeneousEvent &iEvent,
+                      const edm::EventSetup &iSetup,
+                      cuda::stream_t<> &cudaStream) override;
+  void produceGPUCuda(edm::HeterogeneousEvent &iEvent,
+                      const edm::EventSetup &iSetup,
+                      cuda::stream_t<> &cudaStream) override;
+  void produceCPU(edm::HeterogeneousEvent &iEvent,
+                  const edm::EventSetup &iSetup) override;
+
+private:
+  edm::EDGetTokenT<IntermediateHitDoublets> doubletToken_;
+
+  edm::RunningAverage localRA_;
+  CAHitQuadrupletGeneratorGPU GPUGenerator_;
+  CAHitQuadrupletGenerator CPUGenerator_;
+
+  bool emptyRegionDoublets = false;
+  std::unique_ptr<RegionsSeedingHitSets> seedingHitSets_;
+  std::vector<OrderedHitSeeds> ntuplets_;
+};
+
+CAHitNtupletHeterogeneousEDProducer::CAHitNtupletHeterogeneousEDProducer(
+    const edm::ParameterSet &iConfig)
+    : HeterogeneousEDProducer(iConfig),
+      doubletToken_(consumes<IntermediateHitDoublets>(
+          iConfig.getParameter<edm::InputTag>("doublets"))),
+      GPUGenerator_(iConfig, consumesCollector()),
+      CPUGenerator_(iConfig, consumesCollector()) {
+  produces<RegionsSeedingHitSets>();
+}
+
+void CAHitNtupletHeterogeneousEDProducer::fillDescriptions(
+    edm::ConfigurationDescriptions &descriptions) {
+  edm::ParameterSetDescription desc;
+
+  desc.add<edm::InputTag>("doublets", edm::InputTag("hitPairEDProducer"));
+  CAHitQuadrupletGeneratorGPU::fillDescriptions(desc);
+  HeterogeneousEDProducer::fillPSetDescription(desc);
+  auto label = CAHitQuadrupletGeneratorGPU::fillDescriptionsLabel() +
+               std::string("EDProducer");
+  descriptions.add(label, desc);
```

Comment:
Replace with
```cpp
descriptions.add("caHitQuadrupletHeterogeneousEDProducer", desc)
```
(you can also remove `CAHitQuadrupletGeneratorGPU::fillDescriptionsLabel()` if you want)